### PR TITLE
Add `lock` wrapper for Channels.

### PR DESCRIPTION
**Pull request by TommyXR**
`lock(f, lck)` is defined for locks, but not for Channels. I think it's simply a mistake, thus I added `lock(f, ::Channel)` for API uniformity.

### CODE REVIEW

**Review by DilumAluthge — APPROVED** [review]

**Comment by TommyXR** [comment]
Some people on the slack have argued that we should instead remove the `::ReentrantLock` type restriction in the other file.

I don't mind changing the commit to reflect this, but I would like to know what people think is the better option.

**Comment by DilumAluthge** [comment]
I don't feel strongly either way.

**Comment by JeffBezanson** [comment]
Duplicating the code is not good. There are two other methods that use this approach:
```
lock(f, c::GenericCondition) = lock(f, c.lock)
```
That's better since there's no duplication, but in fact we could just remove the `::AbstractLock` restriction and remove the existing extra methods.

More weirdness:
- `lock(::WeakKeyDict)` is not defined but should be
- `unlock(f, c::GenericCondition) = unlock(f, c.lock)` is defined, but there is no similar unlock method for it to call (?)

**Comment by TommyXR** [comment]
> Duplicating the code is not good. There are two other methods that use this approach:
> 
> ```
> lock(f, c::GenericCondition) = lock(f, c.lock)
> ```
> 
> That's better since there's no duplication, but in fact we could just remove the `::AbstractLock` restriction and remove the existing extra methods.
> 
> More weirdness:
> 
>     * `lock(::WeakKeyDict)` is not defined but should be (also `unlock`)
> 
>     * `unlock(f, c::GenericCondition) = unlock(f, c.lock)` is defined, but there is no similar unlock method for it to call (?)

I actually missed the fact that `Channel.cond_take` was actually a condition and that the method existed for `GenericCondition`. Made the change. I also removed `unlock(f, ::GenericCondition)` because after looking for a bit it's definitely a mistake and I added the missing function for `WeakKeyDict`.

File: base/weakkeydict.jl [hunk 1]
```diff
@@ -84,6 +84,7 @@ empty(d::WeakKeyDict, ::Type{K}, ::Type{V}) where {K, V} = WeakKeyDict{K, V}()
 IteratorSize(::Type{<:WeakKeyDict}) = SizeUnknown()
 
 islocked(wkh::WeakKeyDict) = islocked(wkh.lock)
+lock(wkh::WeakKeyDict) = lock(wkh.lock)
```
Comment (JeffBezanson) [inline base/weakkeydict.jl:87]: Also needs `unlock`.

**Comment by JeffBezanson** [comment]
It looks like we are also missing `lock(f, ::IO)`. That makes me think it's better just to remove all but one of the 2-argument `lock` methods, and define `lock(f, l::Any)`.

File: base/weakkeydict.jl (hunk 1 above)
Comment (vtjnash) [reply base/weakkeydict.jl:87]:
```suggestion
unlock(wkh::WeakKeyDict) = unlock(wkh.lock)
```

File: base/weakkeydict.jl [hunk 2]
```diff
@@ -84,6 +84,7 @@ empty(d::WeakKeyDict, ::Type{K}, ::Type{V}) where {K, V} = WeakKeyDict{K, V}()
 IteratorSize(::Type{<:WeakKeyDict}) = SizeUnknown()
 
 islocked(wkh::WeakKeyDict) = islocked(wkh.lock)
+unlock(wkh::WeakKeyDict) = unlock(wkh.lock)
```
Comment (vtjnash) [inline base/weakkeydict.jl:87]:
```suggestion
lock(wkh::WeakKeyDict) = lock(wkh.lock)
unlock(wkh::WeakKeyDict) = unlock(wkh.lock)
```

**Pull request merged**